### ISSUE TITLE
Improve DL4J UI

### ISSUE THREAD

**Issue by AlexDBlack** [deeplearning4j/deeplearning4j#2351]
0.7.0 added a new UI. This issue is to summarize todos and improvements

To-do list for the DL4J UI:

- Support different sessions on the one UI (i.e., client not server side session sellection)
- Add evaluation (i.e., every n>=1 epochs calculate test set accuracy etc)
- ~~More I18N: Chinese, Ukranian, and fixes for other languages~~
- ~~Allow zooming in/out on the model graph~~
- Improve backend efficiency: cache results where appropriate (rather than querying stats storage every time)
- Add NTP (network time protocol) option for times (StatsListener) (most useful for spark/distributed training)
- Known issue: switching session doesn't refresh/update the graph on the model page.
- Fixes for layerwise pretraining (RBM, VAE)
- Use a binary format (bson or similar perhaps?)

Also: fix the key for large networks...
![image](https://cloud.githubusercontent.com/assets/2360237/21840339/1a7b23a0-d831-11e6-8a9a-231e8079fb85.png)

**Comment by thhart** [comment]
1. Sessions: Is it possible to give sessions a name from the code, so you can better navigate between different runs.

2. ++

3. Regarding the languages, I could offer to translate to German but I have my doubts it really makes sense. Your audience really has to speak English, so I don't see the priority.

**Comment by AlexDBlack** [comment]
@thhart fyi it is possible to set session ID via one of the stats listener constructors (but I agree that could be a lot easier)
https://github.com/deeplearning4j/deeplearning4j/blob/0d6c5f38a77f51ca7b800ab2478dad810750d966/deeplearning4j-ui-parent/deeplearning4j-ui-model/src/main/java/org/deeplearning4j/ui/stats/StatsListener.java#L98

**Comment by gks141270** [comment]
In DL4J UI, I have noticed one thing in iterations/score graph,, that values for score and iterations are getting updated only when I move mouse cursor over graph line. It does not remain updated automatically as iterations goes ahead. Same observation is true for other graphs too.

**Comment by AlexDBlack** [comment]
@thhart btw, re: German translation: I'm not sure how large our German userbase is, but I certainly wouldn't turn it down if you wanted to contribute it.
FYI to add a language - it's a case of implementing a few files here:
https://github.com/deeplearning4j/deeplearning4j/tree/master/deeplearning4j-ui-parent/deeplearning4j-play/src/main/resources/dl4j_i18n

The file extension (.en, .ja etc - ISO 639) specifies the language. Otherwise the format of the files themselves is pretty straightforward.

**Comment by treo** [comment]
I expect the German userbase to actually prefer English as the UI language, as it makes it way easier to ask questions. And anyone who is going to use dl4j should have a rather solid understanding of English anyway, as pretty much everything that you have to learn in order to effectively use it is in English anyway.

**Comment by yinyayun** [comment]
I found the page loading slowly, and sometimes even can't open the page

**Comment by agibsonccc** [comment]
@yinyayun Could you give us something a little less vague? We aren't psychic. We have zero ways to reproduce what you're seeing.

**Comment by yinyayun** [comment]
Soory,I will show detail.

dependency:
```xml
<dependency>
        <groupId>org.deeplearning4j</groupId>
        <artifactId>deeplearning4j-ui_2.10</artifactId>
        <version>0.7.0</version>
    </dependency>
```

run the following code on JDK7:
```java
   UIServer uiServer = UIServer.getInstance();
   StatsStorage statsStorage = new InMemoryStatsStorage();
   uiServer.attach(statsStorage);
```

below i use chrome to open http://localhost:9000/train/overview, i found the loading is very slow.
then i use the developer tools of chrome get screenshots:
[screenshots](https://github.com/yinyayun/other/blob/master/ui-pending.bmp)

I don't know why will wait for a resource:http://fonts.googleapis.com/css....

**Comment by yinyayun** [comment]
Soory,I will show detail.

dependency:
```xml
<dependency>
        <groupId>org.deeplearning4j</groupId>
        <artifactId>deeplearning4j-ui_2.10</artifactId>
        <version>0.7.0</version>
    </dependency>
```

run the following code on JDK7:
```java
   UIServer uiServer = UIServer.getInstance();
   StatsStorage statsStorage = new InMemoryStatsStorage();
   uiServer.attach(statsStorage);
```

below i use chrome to open http://localhost:9000/train/overview, i found the loading is very slow.
then i use the developer tools of chrome get screenshots:
![screenshots](https://github.com/yinyayun/other/blob/master/ui-pending.bmp)

I don't know why will wait for a resource:http://fonts.googleapis.com/css....

**Comment by yinyayun** [comment]
Soory,I will show detail:

dependency:
```xml
<dependency>
    <groupId>org.deeplearning4j</groupId>
    <artifactId>deeplearning4j-ui_2.10</artifactId>
    <version>0.7.0</version>
 </dependency>
```

run the following code on JDK7:
```java
UIServer uiServer = UIServer.getInstance();
StatsStorage statsStorage = new InMemoryStatsStorage();
uiServer.attach(statsStorage);
```

below i use chrome to open http://localhost:9000/train/overview, i found the loading is very slow.
then i use the developer tools of chrome get screenshots:
![screenshots](http://img0.ph.126.net/NagwyvO6mX3TPVysMegViA==/6632203561210803111.png)

I don't know why will wait for a resource:http://fonts.googleapis.com/css....

**Comment by yinyayun** [comment]
My machine cannot be connected to the Internet, now I add a record to the host file,content is as follows:
`
127.0.0.1  fonts.googleapis.com
`
and the page load quickly,i know this way not too elegant, have other method to solve it?

**Comment by reidswanson** [comment]
I'm not sure if adding my 2cents on the UI is appropriate as a comment here, but anyway here are a couple of suggestions:
* Enable zooming to parts of the model score vs iteration graph (e.g. similar to [google stock charts](https://www.google.com/finance?q=INDEXDJX%3A.DJI%2CINDEXSP%3A.INX%2CINDEXNASDAQ%3A.IXIC&ei=MyqVWMn2JMXEjAGesYbABw)). It can be very hard to tell visually if you are making progress after a while because the scale is all screwed up from large early score improvements. This could also be useful on some of the other graphs too (e.g., the learning rate if you're using a decay policy).
* And/or enable the option to plot the score vs iteration graph on a log scale
* Allow the user to show/hide the summary line

**Comment by raver119** [comment]
@AlexDBlack it's safe to close now i think?

**Comment by AlexDBlack** [comment]
A lot of those issues are resolved, but not all of them. UI still needs some polishing work, but theer's other issues for that IIRC.
